### PR TITLE
Compatibility with Webpack 5

### PR DESCRIPTION
**Pull request by gauravtiwari**
### 1. Automatic rules

See the approach in: `package/rules/index.js`

The idea is that a rule is only loaded if you have declared the relevant dependency in package.json file. This aligns webpacker with the intention of its creation i.e. to only use it for JS compilation. No need to keep extra deps around. 

There is still some work left to do to ensure errors are more clear and concise. 
Please feel free to add comments.  

### 2. Simpler webpack config 

Change from class-based config files to regular object-based config 
as seen in webpack documentation to make learning and customisation easier 
for everyone.

### 3. And, webpack 5

Obviously, ensuring the latest release is compatible with Webpack 5.0

Hopefully, I will get this ready and merged this week and then we can push out 6.0 release

Supersedes #2797 
See #2757 

## Out of the box compatibility checklist: 

- [x] JS
- [x] CSS
- [x] SASS
- [x] Tailwindcss
- [x] React
- [x] Assets
- [x] SVG

### CODE REVIEW

File: package/environments/base.js [hunk 1]
```diff
@@ -99,82 +84,36 @@ const getBaseConfig = () => new ConfigObject({
...
 
+  plugins: getPlugins(),
+
+  optimization: {
+    splitChunks: {
+      chunks: 'all',
+      name: true
+    },
+    runtimeChunk: 'single'
+  },
```
Comment (rossta) [inline package/environments/base.js:101]: Thinking of leaving `òptimization.splitChunks` on by default? That would be interesting! Perhaps we might consider then swapping out the implementation of `javascript_pack_tag` with that of `javascript_packs_with_chunks_tag` to make that the default behavior.

**Comment by rossta** [comment]
@gauravtiwari Thanks for sharing the work-in-progress. I'll give it a closer look this week.

**Comment by rossta** [comment]
Good timing. Support for Webpack 5 [just released](https://github.com/rails/webpacker/issues/2757#issuecomment-742101856) in webpack-assets-manifest.

https://github.com/webdeveric/webpack-assets-manifest/releases/tag/v5.0.0

**Comment by gauravtiwari** [comment]
thanks @rossta I will get this sorted in the weekend.

**Comment by pedrofurtado** [comment]
Happy to see this merged asap! 🎉 🏆

**Comment by pedrofurtado** [comment]
A suggestion: Change PR name to "Compatibility with Webpack 5" or something similar 🤝 🍻

**Comment by pedrofurtado** [comment]
Any news about this awesome PR? We can help here with testing it in our applications 👍 🏆

**Comment by gauravtiwari** [comment]
Almost there, got some testing left to do which I will finish off tomorrow. 

I might need some help moving installers steps into docs, mainly `yarn adds`

**Comment by gauravtiwari** [comment]
You can try out yourself:

```bash
yarn add https://github.com/rails/webpacker\#automatic-rules
gem 'webpacker', github: 'rails/webpacker', branch: 'automatic-rules'
bundle exec rails webpacker:install
```

```bash
# application.html.erb
javascript_packs_with_chunks_tag('application')
```

**Comment by pedrofurtado** [comment]
Great news! I will test here, using these commands above in your comment @gauravtiwari (thanks for instructions!), to give some feedbacks 🤝

File: package/environments/base.js (hunk 1 above)
Comment (gauravtiwari) [reply package/environments/base.js]: Yeah, I think that makes sense for proper long term caching support (this is recommended in webpack docs)

**Comment by gauravtiwari** [comment]
Hi all,

I have released a new pre-version (6.0):

```
yarn add @rails/webpacker@6.0.0-pre.2
```

Gemfile:

```rb
gem 'webpacker', '~> 6.x'
```

Run:

bundle && bundle exec rails webpacker:install

and install any dependencies (if you are using any). For example, to enable CSS support

```
yarn add css-loader style-loader sass-loader mini-css-extract-plugin
```

To support postcss, add (plus any plugins in postcss.config):

```
yarn add postcss-loader
```

Add to your layouts file (split chunks is now turned on by default):

```erb
    <%= stylesheet_packs_with_chunks_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
    <%= javascript_packs_with_chunks_tag 'application', 'data-turbolinks-track': 'reload' %>
```

There are some breaking changes but hopefully, migration won't be painful :) 

Please help me clean up docs.

**Comment by gauravtiwari** [comment]
Hopefully, we can get 6.0 stable ready just before Christmas 🎄 🎁

File: docs/css.md [hunk 2]
```diff
@@ -298,11 +291,13 @@ yarn add --dev typings-for-css-modules-loader
...
 
 // replace css-loader with typings-for-css-modules-loader
-environment.loaders.get('moduleSass').use = environment.loaders.get('moduleSass').use.map((u) => {
-  if(u.loader == 'css-loader') {
-    return { ...u, loader: 'typings-for-css-modules-loader' };
-  } else {
-    return u;
-  }
-});
+environment.loaders.get('moduleSass').use = environment.loaders
```
Comment (pustovalov) [inline docs/css.md:294]: Hey @gauravtiwari 

should it be `loaders.get('moduleSass').use` ?

```
const { environment, loaders } = require('@rails/webpacker')

console.log('environment', environment)
console.log('loaders', loaders)

environment undefined
loaders [
  {
    test: [
```

Comment (gauravtiwari) [reply docs/css.md:294]: @pustovalov Hey, I think we need to overhaul the documentation. The class-based API is removed in 6.x

I will work on updating docs.

**Comment by pedrofurtado** [comment]
> Hi all,
> 
> I have released a new pre-version (6.0):
> 
> ```
> yarn add @rails/webpacker@6.0.0-pre.2
> ```
> 
> Gemfile:
> 
> ```ruby
> gem 'webpacker', '~> 6.x'
> ```
> 
> Run:
> 
> bundle && bundle exec rails webpacker:install
> 
> and install any dependencies (if you are using any). For example, to enable CSS support
> 
> ```
> yarn add css-loader style-loader sass-loader mini-css-extract-plugin
> ```
> 
> To support postcss, add (plus any plugins in postcss.config):
> 
> ```
> yarn add postcss-loader
> ```
> 
> Add to your layouts file (split chunks is now turned on by default):
> 
> ```
>     <%= stylesheet_packs_with_chunks_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
>     <%= javascript_packs_with_chunks_tag 'application', 'data-turbolinks-track': 'reload' %>
> ```
> 
> There are some breaking changes but hopefully, migration won't be painful :)
> 
> Please help me clean up docs.

Thanks @gauravtiwari ! Once docs are updated, to expose the migration guide from 5.x to 6.x version, we can help with testing 🎉

**Comment by pixeltrix** [comment]
@gauravtiwari the configuration change from this:

``` javascript
process.env.NODE_ENV = process.env.NODE_ENV || 'development'
const { environment } = require('@rails/webpacker')
module.exports = environment.toWebpackConfig()
```

to this:

``` javascript
process.env.NODE_ENV = process.env.NODE_ENV || 'development'
const webpackConfig = require('./base')
module.exports = webpackConfig
```

is going to lead to compilation errors like:

```
[webpack-cli] TypeError: Cannot read property 'toWebpackConfig' of undefined
```

in upgrading apps.

What's our strategy for handling this because currently all the relevant stable branches on Rails are broken (@yahonda [fixed master by locking to the 5.2.1 npm package version][1]). Is it possible to add a deprecation warning on this and still make the old files work?

[1]: https://github.com/rails/rails/commit/437fd8f3c7f2978a94bd142f5d5654cd8ea662f0

**Comment by gauravtiwari** [comment]
Hi Andrew,

Yes, we can do that. Let me add the old API back in but with warning that it will be removed completely in Webpacker 6.1

Is that sound good to you?

**Comment by pixeltrix** [comment]
@gauravtiwari yes, that sounds great - thanks 👍

File: lib/webpacker/helper.rb [hunk 3]
```diff
@@ -148,17 +140,10 @@ def preload_pack_asset(name, **options)
...
-  #   nil
-  #
-  #   # When extract_css is true in webpacker.yml:
   #   <%= stylesheet_pack_tag 'calendar', 'data-turbolinks-track': 'reload' %> # =>
   #   <link rel="stylesheet" media="screen" href="/packs/calendar-1016838bab065ae1e122.css" data-turbolinks-track="reload" />
   def stylesheet_pack_tag(*names, **options)
-    if current_webpacker_instance.config.extract_css?
-      stylesheet_link_tag(*sources_from_manifest_entries(names, type: :stylesheet), **options)
-    end
+    stylesheet_link_tag(*sources_from_manifest_entries(names, type: :stylesheet), **options)
```
Comment (justin808) [inline lib/webpacker/helper.rb:142]: @gauravtiwari @rossta why was this removed? If one is using HMR with React, we don't want to extract the CSS. We want the CSS to come from the JS files so that it can be dynamically loaded in the browser.

File: package/utils/get_style_rule.js [hunk 4]
```diff
@@ -1,43 +1,38 @@
...
+        importLoaders: 2
       }
     },
+    tryPostcss(),
     ...preprocessors
-  ]
-
-  const options = modules ? { include: /\.module\.[a-z]+$/i } : { exclude: /\.module\.[a-z]+$/i }
-
-  if (config.extract_css) {
```
Comment (justin808) [inline package/utils/get_style_rule.js:32]: @gauravtiwari This should be put back. We need this functionality for HMR.

**Review by justin808 — COMMENTED** [review]
@gauravtiwari We can either put back in the extract_css option or we can turn smartly disable extraction when config matches processing matches up `(process.env.WEBPACK_DEV_SERVER && process.env.WEBPACK_DEV_SERVER !== 'undefined')`

Since the rails process also knows if the dev server is running and the value of the `dev_server.hmr`, that can smartly pick whether or not to create the view helper tag or leave it blank.
